File: 76-minimum-window-substring/minimum-window-substring.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        if (s.empty() || t.empty() || s.length() < t.length()) {
            return "";
        }

        vector<int> map(128, 0);
        int count = t.length();
        int l = 0, r = 0, minLen = INT_MAX, i = 0;
        for (char c : t) {
            map[c]++;
        }

        while(r < s.length()) {
            if (map[s[r++]]-- > 0) {
                count--;
            }

            while(count == 0) {
                if (r - l < minLen) {
                    i = l;
                    minLen = r - l;
                }

                if (map[s[l++]]++ == 0) {
                    count++;
                }
            }
        }

        return minLen == INT_MAX ? "" : s.substr(i, minLen);
    }
};

File: 76-minimum-window-substring/minimum-window-substring.java
class Solution {
    public String minWindow(String s, String t) {
        if (s.isEmpty() || t.isEmpty() || s.length() < t.length()) {
            return "";
        }

        HashMap<Character, Integer> window = new HashMap<>();
        HashMap<Character, Integer> need = new HashMap<>();
        int l = 0;
        int r = 0;
        int start = 0;
        int valid = 0;
        int minLen = Integer.MAX_VALUE;

        for(char c : t.toCharArray()) {
            need.put(c, need.getOrDefault(c, 0) + 1);
        }

        while(r < s.length()) {
            char c = s.charAt(r);
            r++;

            if (need.containsKey(c)) {
                window.put(c, window.getOrDefault(c, 0) + 1);

                if (window.get(c).equals(need.get(c))) {
                    valid++;
                }
            }

            while(need.size() == valid) {
                if (r - l < minLen) {
                    start = l;
                    minLen = r - l;
                }

                char d = s.charAt(l);
                l++;
                if (need.containsKey(d)) {
                    if (window.get(d).equals(need.get(d))) {
                        valid--;
                    }
                    window.put(d, window.get(d) - 1);
                }
            }
        }

        return minLen == Integer.MAX_VALUE ? "" : s.substring(start, start+minLen);
    }
}